README.md
# Check your application for broken links

[![GitHub Tests Action Status](https://github.com/lobotomised/laravel-autocrawler/actions/workflows/run-test.yml/badge.svg)](https://github.com/lobotomised/laravel-autocrawler/actions/workflows/run-test.yml)
[![Total Downloads](https://img.shields.io/packagist/dt/lobotomised/laravel-autocrawler.svg?style=flat-square)](https://packagist.org/packages/lobotomised/laravel-autocrawler)
[![Latest Stable Version](https://img.shields.io/packagist/v/lobotomised/laravel-autocrawler)](https://packagist.org/packages/lobotomised/laravel-autocrawler)
[![License](https://img.shields.io/packagist/l/lobotomised/laravel-autocrawler)](https://packagist.org/packages/lobotomised/laravel-autocrawler)

Using this package you can check if your application have broken links.

```bash
php artisan crawl
200 OK - http://myapp.test/ 
200 OK - http://myapp.test/login found on http://myapp.test/
200 OK - http://myapp.test/register found on http://myapp.test/
301 301 Moved Permanently - http://myapp.test/homepage found on http://myapp.test/register
404 Not Found - http://myapp.test/brokenlink found on http://myapp.test/register
200 OK - http://myapp.test/features found on http://myapp.test/


Crawl finished

Results:
Status 200: 4 founds
Status 301: 1 found
Status 404: 1 found
```

## Installation
This package can be installed via Composer:

``` bash
composer require --dev lobotomised/laravel-autocrawler
```

When crawling your site, it will automatically detect the url your application is using. If instead it scan http://localhost, check in your .env you properly configure the APP_URL variable
``` dotenv
APP_URL="http://myapp.test"
``` 

## Usage

### Crawl a specific url
By default, the crawler will crawl the URL from your current laravel installation. You can force the url with the  `--url` option:
```bash
php artisan crawl --url=http://myapp.test/my-page
``` 
### Concurrent connection
The crawler run with 10 concurrent connections to speed up the crawling process. You can change that by passing the `--concurrency` option:
```bash
php artisan crawl --concurrency=5
```
### Timeout
The request timeout is by default 30 seconds. Use the `--timeout` to change this value
```bash
php artisan crawl --timeout=10
```
### Ignore robots.txt
By default, the crawler respect the robots.txt. These rules can be ignored with the `--ignore-robots` option:
```bash
php artisan crawl --ignore-robots
```
### External link
When the crawler find an external link, it will check this link. It can be deactivated with the `--ignore-external-links` option:
```bash
php artisan crawl --ignore-external-links
```
### Log non-2xx or non-3xx status code
By default, the crawler will only in your console. You can log all non-2xx or non 3xx status code to a file with the `--output` option. Result will be store in `storage/autocrawler/output.txt`
```bash
php artisan crawl --output
```
The output.txt will look like that:
```
403 Forbidden - http://myapp.test/dashboard found on http://myapp.test/home
404 Not Found - http://myapp.test/brokenlink found on http://myapp.test/register
```
### Fail when non-2xx or non-3xx are found
By default, the command exit codes is 0. You can change it to 1 to indicate that the command has failed with the `--fail-on-error`
```bash
php artisan crawl --fail-on-error
```
### Launch the robot interactively
Eventually, you may configure the crawler interactively by using the `--interactive` option:
```bash
php artisan crawl --interactive
```

## Working with GitHub actions
To execute the crawler you first need to start a web server. You can choose to install apache or nginx. 
Here is an example using the php build-in webserver

If the crawl found some non-2xx or non-3xx response, the action will fail, and the result will be store as an artifacts of the Action.

```
steps:
  - uses: actions/checkout@v3
  - name: Prepare The Environment
    run: cp .env.example .env
  - name: Install Composer Dependencies
    run: composer install
  - name: Generate Application Key
    run: php artisan key:generate
  - name: Install npm Dependencies
    run: npm ci
  - name: Compile assets
    run: npm run build

  - name: Start php build-in webserver
    run: (php artisan serve &) || /bin/true

  - name: Crawl website
    run: php artisan crawl --url=http://localhost:8000/ --fail-on-error --output
  
  - name: Upload artifacts
    if: failure()
    uses: actions/upload-artifact@master
    with:
      name: Autocrawler
      path: ./storage/autocrawler
``` 

## Documentation
All commands and informations are available with the command:
```bash
php artisan crawl --help
```

## Alternatives
This package is heavily inspire by [spatie/http-status-check](https://github.com/spatie/http-status-check), but instead of being a project dependency, it is a global installation

## Testing

First we need to start the included node http server in a separate terminal.
```bash
make start
```
Then to run the tests:
```bash
make test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-autocrawler` will be documented in this file.

## 1.1.0 - 2023-02-12

**Full Changelog**: https://github.com/lobotomised/laravel-autocrawler/compare/1.0.1...1.1.0

## 1.0.0 - 2022-08-18

Initial release

## 0.0.4 - 2022-08-17

**Full Changelog**: https://github.com/lobotomised/laravel-autocrawler/compare/0.0.3...0.0.4

## 0.0.3 - 2022-08-17

**Full Changelog**: https://github.com/lobotomised/laravel-autocrawler/compare/0.0.2...0.0.3


tests/Pest.php
<?php

declare(strict_types=1);

use Illuminate\Filesystem\Filesystem;
use Lobotomised\Autocrawler\Tests\TestCase;

uses(TestCase::class)
    ->in(__DIR__);

expect()->extend('toBeOne', function () {
    return $this->toBe(1);
});

function expectToLog(): void
{
    $result = (new Filesystem())->isFile(storage_path('autocrawler/output.txt'));

    expect($result)->toBeTrue();
}

function expectToNotLog(): void
{
    $result = (new Filesystem())->isFile(storage_path('autocrawler/output.txt'));

    expect($result)->toBeFalse();
}


tests/TestCase.php
<?php

declare(strict_types=1);

namespace Lobotomised\Autocrawler\Tests;

use Lobotomised\Autocrawler\AutocrawlerServiceProvider;

class TestCase extends \Orchestra\Testbench\TestCase
{
    protected function getPackageProviders($app)
    {
        return [
            AutocrawlerServiceProvider::class,
        ];
    }
}


tests/server/server.js
"use strict";

const app = require('express')();
const morgan = require('morgan')

app.use(morgan('combined'))

app.get('/200', function(req, res) {
    res.end('<html><body><p>ok</p></body></html>');
});

app.get('/301', function(req, res) {
    res.redirect(301,'/ok');
});

app.get('/404', function (req, res) {
    res.status(404).end();
})

app.get('/page1', function(req, res) {
    res.end('<html><body><a href="/page2">foo</a></body></html>');
});
app.get('/page2', function(req, res) {
    res.end('<html><body><p>ok</p></body></html>');
});

let server = app.listen(8080, function () {
    const host = 'localhost';
    const port = server.address().port;

    console.log('Testing server listening at http://%s:%s', host, port);
});

src/Crawler.php
<?php

declare(strict_types=1);

namespace Lobotomised\Autocrawler;

use GuzzleHttp\Psr7\Uri;
use GuzzleHttp\RequestOptions;
use Psr\Http\Message\UriInterface;
use Spatie\Crawler\Crawler as SpatieCrawler;
use Spatie\Crawler\CrawlProfiles\CrawlAllUrls;
use Spatie\Crawler\CrawlProfiles\CrawlInternalUrls;
use Spatie\Crawler\CrawlProfiles\CrawlProfile;
use Symfony\Component\Console\Output\OutputInterface;

class Crawler
{
    private OutputInterface $consoleOutput;

    private UriInterface $baseUrl;

    private int $concurrency = 10;

    private bool $ignore_robots = false;

    private CrawlProfile $crawlProfile;

    private bool $shouldOutput = false;

    /**
     * @var array <string, bool|int|array<string, bool>>
     */
    protected static array $defaultClientOptions = [
        RequestOptions::COOKIES => true,
        RequestOptions::CONNECT_TIMEOUT => 10,
        RequestOptions::TIMEOUT => 10,
        RequestOptions::ALLOW_REDIRECTS => [
            'track_redirects' => true,
        ],
    ];

    public function setConsoleOutput(OutputInterface $output): self
    {
        $this->consoleOutput = $output;

        return $this;
    }

    public function setUrl(UriInterface|string $url): self
    {
        if (! $url instanceof UriInterface) {
            $url = new Uri($url);
        }

        $this->baseUrl = $url;

        return $this;
    }

    public function setConcurrency(int $concurrency): self
    {
        $this->concurrency = $concurrency;

        return $this;
    }

    public function setTimeout(int $timeout): self
    {
        self::$defaultClientOptions[RequestOptions::TIMEOUT] = $timeout;

        return $this;
    }

    public function setIgnoreRobots(bool $ignore_robots): self
    {
        $this->ignore_robots = $ignore_robots;

        return $this;
    }

    public function setCrawlProfile(bool $crawlProfile): self
    {
        if ($crawlProfile) {
            $this->crawlProfile = new CrawlInternalUrls($this->baseUrl);

            return $this;
        }

        $this->crawlProfile = new CrawlAllUrls();

        return $this;
    }

    public function startCrawling(): bool
    {
        $observer = new CrawlerObserver($this->consoleOutput);

        if ($this->shouldOutput) {
            $observer->shouldOutput($this->shouldOutput);
        }

        $crawler = SpatieCrawler::create([])
            ->setCrawlObserver($observer)
            ->setConcurrency($this->concurrency)
            ->setCrawlProfile($this->crawlProfile);

        $this->ignore_robots
            ? $crawler->ignoreRobots()
            : $crawler->respectRobots();

        $crawler->startCrawling($this->baseUrl);

        foreach ($observer->result() as $code => $urls) {
            if ($code >= 400 || $code === 0) {
                return false;
            }
        }

        return true;
    }

    public function shouldOutput(bool $output): self
    {
        $this->shouldOutput = $output;

        return $this;
    }
}


src/CrawlerObserver.php
<?php

declare(strict_types=1);

namespace Lobotomised\Autocrawler;

use GuzzleHttp\Exception\RequestException;
use Illuminate\Filesystem\Filesystem;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\UriInterface;
use Spatie\Crawler\CrawlObservers\CrawlObserver;
use Symfony\Component\Console\Output\OutputInterface;

class CrawlerObserver extends CrawlObserver
{
    /** @var array<int, array<int, UriInterface>> */
    protected array $crawledUrls = [];

    private Filesystem $files;

    private bool $shouldOutput = false;

    private const DIRECTORY = 'autocrawler';

    public function __construct(private OutputInterface $consoleOutput)
    {
        $this->files = new Filesystem();
        $this->createLogDirectory();
    }

    public function shouldOutput(bool $shouldOutput): void
    {
        $this->shouldOutput = $shouldOutput;
    }

    public function crawled(UriInterface $url, ResponseInterface $response, ?UriInterface $foundOnUrl = null): void
    {
        $this->addResult($url, $foundOnUrl, $response->getStatusCode(), $response->getReasonPhrase());
    }

    public function crawlFailed(UriInterface $url, RequestException $requestException, ?UriInterface $foundOnUrl = null): void
    {
        $response = $requestException->getResponse();

        $this->addResult($url, $foundOnUrl, $response?->getStatusCode() ?? 0, $response?->getReasonPhrase() ?? '');
    }

    public function finishedCrawling(): void
    {
        $this->consoleOutput->writeln("\n<info>Crawl finished</info>");
        $this->consoleOutput->writeln("\n<info>Results:</info>");

        foreach ($this->crawledUrls as $code => $urls) {
            $count = count($urls);
            $txt = $count > 1 ? 'founds' : 'found';
            $colorTag = $this->getColorTag($code);

            $this->consoleOutput->writeln("<$colorTag>Status $code: $count $txt</$colorTag>");
        }
    }

    /**
     * @return array<int, array<int, UriInterface>>
     */
    public function result(): array
    {
        return $this->crawledUrls;
    }

    private function addResult(UriInterface $url, ?UriInterface $foundOnUrl, int $code, string $reason): void
    {
        if (isset($this->crawledUrls[$code]) && in_array($url, $this->crawledUrls[$code])) {
            return;
        }

        $this->crawledUrls[$code][] = $url;

        $colorTag = $this->getColorTag($code);

        $date = date('Y-m-d H:i:s');

        $message = "$code $reason - $url";
        if ($foundOnUrl) {
            $message .= " found on $foundOnUrl";
        }

        $this->consoleOutput->writeln("<$colorTag> [$date] $message</$colorTag>");

        if ($this->shouldOutput && $colorTag === 'error') {
            $this->log($message);
        }
    }

    private function getColorTag(int $code): string
    {
        if (str_starts_with((string) $code, '2')) {
            return 'info';
        }

        if (str_starts_with((string) $code, '3')) {
            return 'comment';
        }

        return 'error';
    }

    private function createLogDirectory(): void
    {
        $dir_path = storage_path(self::DIRECTORY);

        if (! $this->files->isDirectory($dir_path)) {
            if ($this->files->makeDirectory($dir_path)) {
                $this->files->put($dir_path.DIRECTORY_SEPARATOR.'.gitignore', "*\n!.gitignore\n");
            } else {
                throw new \Exception("Cannot create directory 'self::DIRECTORY'");
            }
        }

        $this->files->delete($dir_path.DIRECTORY_SEPARATOR.'output.txt');
    }

    private function log(string $message): void
    {
        $this->files->append(
            storage_path(self::DIRECTORY).DIRECTORY_SEPARATOR.'output.txt',
            $message.PHP_EOL
        );
    }
}


src/Commands/CrawlCommand.php
<?php

declare(strict_types=1);

namespace Lobotomised\Autocrawl\Commands;

use Illuminate\Console\Command;
use Lobotomised\Autocrawl\Crawler;

class CrawlCommand extends Command
{
    public $signature = 'crawl  {--u|url= : Url to crawl. Default your current laravel app}
                                {--c|concurrency=10 : Number of concurrent connections}
                                {--t|timeout=30 : Max duration of a request}
                                {--r|ignore-robots : Should ignore rules in the /robots.txt} 
                                {--e|ignore-external-links : Do not follow external links}
                                {--o|output= : Write all non-2** ou non-3** to a file}
                                {--i|interactive : Interactive ask your for the options}
                                {--fail-on-error: Command will fail if HTTP 400 or 500 codes are detected}';

    public $description = 'Crawl your own website and search failing status code';

    public function __construct(private Crawler $crawler)
    {
        parent::__construct();
    }

    public function handle(): int
    {
        $baseUrl = $this->option('url') ?: config('app.url');

        if($this->option('interactive')) {
            $concurrency = (int) $this->ask('Number of concurrent connections ?', $this->option('concurrency'));
            $timeout = (int) $this->ask('Max duration of a request ?', $this->option('timeout'));
            $baseUrl = $this->ask('Url to crawl ?', $baseUrl);
            $output = $this->ask('Filename where we store all non-2** ou non-3** response. Leave blank to not write to a file', false);
            $ignore_robots = $this->confirm('Should ignore rules in the /robots.txt', $this->option('ignore-robots'));
            $ignore_external_links = $this->confirm('Ignore external links ?', $this->option('ignore-external-links'));
        }

        $this->crawler->setUrl($baseUrl)
            ->setOutput($this->getOutput())
            ->setOutputToFile($output ?? $this->option('output'))
            ->setConcurrency($concurrency ?? (int) $this->option('concurrency'))
            ->setTimeout($timeout ?? (int) $this->option('timeout'))
            ->setIgnoreRobots($ignore_robots ?? $this->option('ignore-robots'))
            ->setCrawlProfile($ignore_external_links ?? $this->option('ignore-external-links'))
            ->startCrawling();

        return 0;
    }
}

src/Commands/CrawlerCommand.php
<?php

declare(strict_types=1);

namespace Lobotomised\Autocrawler\Commands;

use Illuminate\Console\Command;
use Lobotomised\Autocrawler\Crawler;

class CrawlerCommand extends Command
{
    public $signature = 'crawl  {--u|url= : Url to crawl. Default your current laravel app}
                                {--c|concurrency=10 : Number of concurrent connections}
                                {--t|timeout=30 : Max duration of a request}
                                {--r|ignore-robots : Should ignore rules in the /robots.txt} 
                                {--e|ignore-external-links : Do not follow external links}
                                {--o|output : Write all non-2** ou non-3** to storage/autocrawler/output.txt}
                                {--i|interactive : Interactive ask your for the options}
                                {--fail-on-error : Command will fail if HTTP 400 or 500 codes are detected}';

    public $description = 'Crawl your own website and search failing status code';

    public function __construct(private Crawler $crawler)
    {
        parent::__construct();
    }

    public function handle(): int
    {
        $baseUrl = $this->option('url')
            ? $this->getOption('url')
            : config('app.url');

        if ($this->option('interactive')) {
            $concurrency = (int) $this->ask('Number of concurrent connections ?', $this->getOption('concurrency'));
            $timeout = (int) $this->ask('Max duration of a request in seconds ?', $this->getOption('timeout'));
            $baseUrl = $this->ask('Url to crawl ?', $baseUrl);
            $output = $this->confirm('Write all non-2** ou non-3** to storage/autocrawler/output.txt', (bool) $this->option('output'));
            $ignore_robots = $this->confirm('Should ignore rules in the /robots.txt', (bool) $this->option('ignore-robots'));
            $ignore_external_links = $this->confirm('Ignore external links ?', (bool) $this->option('ignore-external-links'));
        }

        $result = $this->crawler->setUrl($baseUrl)
            ->setConsoleOutput($this->getOutput())
            ->shouldOutput($output ?? (bool) $this->option('output'))
            ->setConcurrency($concurrency ?? (int) $this->option('concurrency'))
            ->setTimeout($timeout ?? (int) $this->option('timeout'))
            ->setIgnoreRobots($ignore_robots ?? (bool) $this->option('ignore-robots'))
            ->setCrawlProfile($ignore_external_links ?? (bool) $this->option('ignore-external-links'))
            ->startCrawling();

        if ($this->option('fail-on-error') && $result === false) {
            return self::FAILURE;
        }

        return self::SUCCESS;
    }

    private function getOption(string $option): ?string
    {
        $value = $this->option($option);

        if (is_string($value) || $value === null) {
            return $value;
        }

        return null;
    }
}
